File: implementations/python/provekit-lift-py-tests/tests/test_cicp_conformance.py
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import pytest

from provekit_lift_py_tests.canonicalizer import (
    Value,
    jcs_hash,
    varr,
    vbool,
    vint,
    vnull,
    vobj,
    vstr,
)


REPO_ROOT = Path(__file__).resolve().parents[4]
CICP_VECTOR_DIR = REPO_ROOT / "protocol" / "conformance" / "cicp"


def _json_to_value(value: Any) -> Value:
    if value is None:
        return vnull()
    if isinstance(value, bool):
        return vbool(value)
    if isinstance(value, int):
        return vint(value)
    if isinstance(value, str):
        return vstr(value)
    if isinstance(value, list):
        return varr([_json_to_value(item) for item in value])
    if isinstance(value, dict):
        return vobj([(key, _json_to_value(item)) for key, item in value.items()])
    raise TypeError(f"unsupported JSON value in CICP vector: {value!r}")


def _load_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def _assert_no_missing_blast_radius_input_cids(body: dict[str, Any]) -> None:
    input_cids = set(body["inputCids"])
    required_cids = {
        body["protocolCatalogCid"],
        body["jobDefinitionCid"],
        body["commandCid"],
        body["runnerIdentityCid"],
        body["sourceClosureCid"],
        body["policyCid"],
        *body["toolchainCids"],
        *body["lockfileCids"],
        *body["generatedInputCids"],
        *body["fixtureCids"],
        *body["relevantSpecCids"],
    }

    missing = sorted(required_cids - input_cids)
    if missing:
        raise ValueError(f"inputCids missing required CID: {missing[0]}")


def test_cicp_passing_vectors_match_pinned_cids():
    catalog = _load_json(CICP_VECTOR_DIR / "vectors.json")

    for vector in catalog["vectors"]:
        if not vector["shouldPass"]:
            continue

        body = _load_json(CICP_VECTOR_DIR / vector["body"])

        assert jcs_hash(_json_to_value(body)) == vector["expectedCid"]


def test_cicp_invalid_blast_radius_vector_fails_closed_on_missing_input_cid():
    catalog = _load_json(CICP_VECTOR_DIR / "vectors.json")
    vector = next(v for v in catalog["vectors"] if not v["shouldPass"])
    body = _load_json(CICP_VECTOR_DIR / vector["body"])

    with pytest.raises(ValueError, match=vector["errorContains"]):
        _assert_no_missing_blast_radius_input_cids(body)

File: implementations/ruby/test/test_cicp_conformance.rb
# SPDX-License-Identifier: Apache-2.0
#
# CICP language-kit conformance for protocol/conformance/cicp golden vectors.

require "json"
require "minitest/autorun"
require_relative "../lib/provekit/blake3"
require_relative "../lib/provekit/ir"

class TestCicpConformance < Minitest::Test
  ROOT = File.expand_path("../../..", __dir__)
  CICP_DIR = File.join(ROOT, "protocol", "conformance", "cicp")
  VECTORS = JSON.parse(File.read(File.join(CICP_DIR, "vectors.json")))

  def test_passing_vectors_match_blake3_512_of_jcs_body
    passing_vectors.each do |vector|
      body = read_body(vector)
      canonical = Provekit::IR::Jcs.encode(body)

      assert_equal(
        vector.fetch("expectedCid"),
        Provekit::Blake3.hex(canonical),
        "#{vector.fetch("name")} CID must match BLAKE3-512(JCS(body))",
      )
    end
  end

  def test_invalid_blast_radius_fails_closed_on_missing_input_cids_dependency
    vector = VECTORS.fetch("vectors").find { |v| v.fetch("shouldPass") == false }
    error = assert_raises(ArgumentError) { validate_cicp_body(read_body(vector)) }

    assert_includes error.message, vector.fetch("errorContains")
  end

  private

  def passing_vectors
    VECTORS.fetch("vectors").select { |vector| vector.fetch("shouldPass") }
  end

  def read_body(vector)
    JSON.parse(File.read(File.join(CICP_DIR, vector.fetch("body"))))
  end

  def validate_cicp_body(body)
    kind = body.fetch("kind")
    if kind == "CIBlastRadius"
      validate_blast_radius_input_closure(body)
    else
      true
    end
  end

  def validate_blast_radius_input_closure(body)
    input_cids = body.fetch("inputCids")
    required_cids = [
      body.fetch("protocolCatalogCid"),
      body.fetch("jobDefinitionCid"),
      body.fetch("commandCid"),
      body.fetch("runnerIdentityCid"),
      body.fetch("sourceClosureCid"),
      *body.fetch("toolchainCids"),
      *body.fetch("lockfileCids"),
      *body.fetch("generatedInputCids"),
      *body.fetch("fixtureCids"),
      *body.fetch("relevantSpecCids"),
    ]

    missing = required_cids.find { |cid| !input_cids.include?(cid) }
    raise ArgumentError, "inputCids missing required CID: #{missing}" if missing

    true
  end
end
